File: sdks/js_sdk/src/runtime/negotiationEvents.ts
export type NegotiationEventKind = 'open' | 'policy' | 'propose' | 'counter' | 'evaluate' | 'decide' | 'abort' | { unknown: string };

export interface NegotiationEventBase { kind: string; ts?: string }

export interface OpenEvent extends NegotiationEventBase { kind: 'open'; topic: string; corr: string }
export interface PolicyEvent extends NegotiationEventBase { kind: 'policy'; negotiation_id: string; profile?: string; policy: any }
export interface PayloadEventBase extends NegotiationEventBase { from: string; ct: string; payload_b64?: string }
export interface ProposeEvent extends PayloadEventBase { kind: 'propose' }
export interface CounterEvent extends PayloadEventBase { kind: 'counter' }
export interface EvaluateEvent extends NegotiationEventBase { kind: 'evaluate'; from: string; score?: number; notes?: string }
export interface DecideEvent extends NegotiationEventBase { kind: 'decide'; by: string; ct: string; result_b64?: string }
export interface AbortEvent extends NegotiationEventBase { kind: 'abort'; reason?: string }

export type NegotiationEvent = OpenEvent | PolicyEvent | ProposeEvent | CounterEvent | EvaluateEvent | DecideEvent | AbortEvent | NegotiationEventBase;

export function parseNegotiationEvent(json: string): NegotiationEvent {
  try {
    const obj = JSON.parse(json);
    if (!obj || typeof obj.kind !== 'string') return { kind: 'unknown', ...(obj || {}) } as any;
    return obj as NegotiationEvent;
  } catch {
    return { kind: 'unknown', ts: undefined } as any;
  }
}

export function decodeBase64(b64?: string): Uint8Array | undefined {
  if (!b64) return undefined;
  return new Uint8Array(Buffer.from(b64, 'base64'));
}

File: sdks/rust_sdk/src/negotiation_events.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum NegotiationEvent {
    #[serde(rename = "open")]
    Open {
        ts: Option<String>,
        topic: String,
        corr: String,
    },
    #[serde(rename = "policy")]
    Policy {
        ts: Option<String>,
        negotiation_id: String,
        profile: Option<String>,
        policy: serde_json::Value,
    },
    #[serde(rename = "propose")]
    Propose {
        ts: Option<String>,
        from: String,
        ct: String,
        payload_b64: Option<String>,
    },
    #[serde(rename = "counter")]
    Counter {
        ts: Option<String>,
        from: String,
        ct: String,
        payload_b64: Option<String>,
    },
    #[serde(rename = "evaluate")]
    Evaluate {
        ts: Option<String>,
        from: String,
        score: Option<f64>,
        notes: Option<String>,
    },
    #[serde(rename = "decide")]
    Decide {
        ts: Option<String>,
        by: String,
        ct: String,
        result_b64: Option<String>,
    },
    #[serde(rename = "abort")]
    Abort {
        ts: Option<String>,
        reason: Option<String>,
    },
    #[serde(other)]
    Unknown,
}

pub fn parse_negotiation_event(raw: &[u8]) -> Option<NegotiationEvent> {
    serde_json::from_slice::<NegotiationEvent>(raw).ok()
}

pub fn decode_b64(s: &Option<String>) -> Option<Vec<u8>> {
    use base64::engine::general_purpose::STANDARD;
    use base64::Engine;
    s.as_ref().and_then(|b| STANDARD.decode(b).ok())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_policy_event() {
        let json = r#"{"kind":"policy","negotiation_id":"neg-1","policy":{"max_rounds":4}}"#;
        let evt = parse_negotiation_event(json.as_bytes()).unwrap();
        match evt {
            NegotiationEvent::Policy { negotiation_id, .. } => assert_eq!(negotiation_id, "neg-1"),
            _ => panic!("expected policy"),
        }
    }

    #[test]
    fn decode_payload_b64() {
        let s = Some("SGk=".to_string());
        let bytes = decode_b64(&s).unwrap();
        assert_eq!(String::from_utf8(bytes).unwrap(), "Hi");
    }
}
